controllers/issuer_controller.go

```diff
@@ -69,51 +76,58 @@ type IssuerReconciler struct {
 	PostSetupWithManager func(context.Context, schema.GroupVersionKind, ctrl.Manager, controller.Controller) error
 }
 
-func (r *IssuerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, returnedError error) {
+func (r *IssuerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	logger := log.FromContext(ctx).WithName("Reconcile")
 
 	logger.V(2).Info("Starting reconcile loop", "name", req.Name, "namespace", req.Namespace)
 
-	result, issuerStatusPatch, returnedError := r.reconcileStatusPatch(logger, ctx, req)
-	logger.V(2).Info("Got StatusPatch result", "result", result, "patch", issuerStatusPatch, "error", returnedError)
-	if issuerStatusPatch != nil {
-		cr, patch, err := ssaclient.GenerateIssuerStatusPatch(r.ForObject, req.Name, req.Namespace, issuerStatusPatch)
-		if err != nil {
-			return ctrl.Result{}, utilerrors.NewAggregate([]error{err, returnedError})
-		}
+	result, issuerStatusPatch, reconcileError := r.reconcileStatusPatch(logger, ctx, req)
+
+	if issuerStatusPatch == nil {
+		logger.V(2).Info("Got nil StatusPatch result", "result", result, "error", reconcileError)
+		return ctrl.Result{}, reconcileError
+	}
+
+	obj, patch, err := ssaclient.GenerateIssuerStatusPatch(r.ForObject, req.Name, req.Namespace, issuerStatusPatch)
+	if err != nil {
+		return ctrl.Result{}, utilerrors.NewAggregate([]error{err, reconcileError}) // requeue with backoff
+	}
 
-		if err := r.Client.Status().Patch(ctx, cr, patch, &client.SubResourcePatchOptions{
-			PatchOptions: client.PatchOptions{
-				FieldManager: r.FieldOwner,
-				Force:        ptr.To(true),
-			},
-		}); err != nil {
-			if err := client.IgnoreNotFound(err); err != nil {
-				return ctrl.Result{}, utilerrors.NewAggregate([]error{err, returnedError})
-			}
-			logger.V(1).Info("Not found. Ignoring.")
+	logger.V(2).Info("Got StatusPatch result", "result", result, "error", reconcileError, "patch", patch)
+
+	if err := r.Client.Status().Patch(ctx, obj, patch, &client.SubResourcePatchOptions{
+		PatchOptions: client.PatchOptions{
+			FieldManager: r.FieldOwner,
+			Force:        ptr.To(true),
+		},
+	}); err != nil {
+		if err := client.IgnoreNotFound(err); err != nil {
+			return ctrl.Result{}, utilerrors.NewAggregate([]error{err, reconcileError}) // requeue with backoff
 		}
+
+		logger.V(1).Info("Issuer not found. Ignoring.")
+		return ctrl.Result{}, nil // done
 	}
 
-	return result, returnedError
+	return result, reconcileError
 }
 
 func (r *IssuerReconciler) reconcileStatusPatch(
```

Comment:
**Request:** Please explain when each of the returned values can be `nil` in the func `reconcileStatusPatch`.

From what I understand, the only possible cases are:

- `issuerStatusPatch == nil, reconcileError != nil`: happy case
- `issuerStatusPatch == nil, reconcileError == nil`: ??? is it possible to have this case?
- `issuerStatusPatch == nil, reconcileError != nil`: unhappy case

Reply:
I tried to explain this here: https://github.com/cert-manager/issuer-lib/pull/42/files#diff-4baeeaeddf38ed0df2790cd1e77c402812e8cc07f759ca675a83d271b013078dR110-R113

Reply:
My confusion came from the fact that I though the function's purpose was to return `result` and `issuerStatusPatch`, and that it would return an `error` if it failed. The func's description doesn't emphazise on explaining why the returned `error` doesn't mean that `result` or `issuerStatusPatch` weren't produced.